### PR TITLE
Stopping power calculation

### PR DESCRIPTION
**Pull request by Mexolius**

### CODE REVIEW

File: src/Backend/WASMWrapper.js [hunk 1]
```diff
@@ -1,10 +1,11 @@
+import TraceFactory from './TraceFacotry.js';
```
Comment (grzanka) [inline src/Backend/WASMWrapper.js:1]: Should it be Factory, not Facotry ?

File: src/Backend/WASMWrapper.js [hunk 2]
```diff
@@ -79,10 +80,56 @@ export default class WASMWrapper {
...
+    }
+
+    async getTraceByPoints(program, ion, material, points) {
+        const wasm = await this.wasm()
+        const start = wasm.ccall('dedx_get_min_energy', 'number', ['number','number'],[program,ion])
+        const end = wasm.ccall('dedx_get_max_energy', 'number', ['number','number'],[program,ion])
+
+        const xs = TraceFactory.getXAxisByPoints(start,end,points)
+
+        const stepFunction = wasm.cwrap('dedx_get_simple_stp','number',['number', 'number', 'number', 'number'])
```
Comment (grzanka) [inline src/Backend/WASMWrapper.js:97]: Are this functions already implemented and merged into mail libdedx ?

File: .gitignore [hunk 3]
```diff
@@ -26,3 +26,4 @@ src/Backend/weblibdedx.js
 .gitignore
 public/weblibdedx.wasm
 .vscode/settings.json
+public/weblibdedx.data
```
Comment (grzanka) [inline .gitignore:29]: Do we need `public/weblibdedx.data` for correct execution of the GUI ?
If yes, then from where can we get it ?
Can we add it to github actions artifacts, just like webassembly products, i.e:
image

Comment (Mexolius) [reply .gitignore:29]: Oh, forgot about the workflow - It's from the same place the weblibdedx.wasm is comming from.

File: src/Backend/WASMWrapper.js (hunk 2 above)
Comment (Mexolius) [reply src/Backend/WASMWrapper.js:97]: Yes, it's one of the original functions

Comment (Mexolius) [reply src/Backend/WASMWrapper.js:97]: It will still be changed after we introduce the new function working on whole arrays of data

File: .gitignore (hunk 3 above)
Comment (Mexolius) [reply .gitignore:29]: Turns out it's not needed - everything is already compiled in the .js file. It isn't the most efficient setup but the --preload doesn't seem for work with react to well. It shouldn't be that bad as the overall size of the data is ~~3MB counting with spare .h and .py files which will be removed from that folder one day

File: src/Backend/WASMWrapper.js (hunk 2 above)
Comment (grzanka) [reply src/Backend/WASMWrapper.js:97]: ok, got it

File: build_wasm.sh [hunk 4]
```diff
@@ -20,18 +17,21 @@ FUNCTIONS='['
...
+FUNCTIONS+='_dedx_get_max_energy,'
+FUNCTIONS+='_dedx_get_simple_stp,'
 FUNCTIONS+='_malloc,'
 FUNCTIONS+='_free'
 
-
 FUNCTIONS+=']'
 
-emcc libdedx.a -o $PROJECT_NAME.js -s EXPORTED_FUNCTIONS="$FUNCTIONS" -s ENVIRONMENT='web' -s USE_ES6_IMPORT_META=0 -s EXPORT_ES6=1 -s MODULARIZE=1 -s WASM=1 -s EXPORTED_RUNTIME_METHODS=["ccall","cwrap","UTF8ToString"]
+# remember to use --embed-file instead of --preload-file for web based development
```
Comment (grzanka) [inline build_wasm.sh:31]: explain why we should use --embed-file

File: build_wasm.sh [hunk 5]
```diff
@@ -20,18 +17,21 @@ FUNCTIONS='['
...
+emcc libdedx.a -o $PROJECT_NAME.js -s EXPORTED_FUNCTIONS="$FUNCTIONS" -s ENVIRONMENT='web' -s USE_ES6_IMPORT_META=0 -s EXPORT_ES6=1 -s MODULARIZE=1 -s WASM=1 -s EXPORTED_RUNTIME_METHODS=["ccall","cwrap","UTF8ToString"] --embed-file ../../libdedx/data@data/
```
Comment (grzanka) [inline build_wasm.sh:32]: what is that path `../../libdedx/data@data/` ?
Why `@` inside ?

Does it mean mount `../../libdedx/data` inside `wasm` into `data` directory ?

if yes, then explain it in comments

File: .gitignore (hunk 3 above)
Comment (grzanka) [reply .gitignore:29]: if we don't need that, then remove it from `.gitignore`

**Comment by grzanka** [comment]
I've made some tests with GitHub actions artefacts from:
https://github.com/APTG/dedx_web/actions/runs/1508778882
image

and the app is crashing:

image

File: build_wasm.sh (hunk 4 above)
Comment (Mexolius) [reply build_wasm.sh:31]: Embed- as opposed to preload - embeds the data into the output .js file. They both should have the same effect - the data should be available on the site but the preload option requires the generated .js file to load the file when it sets the module up. This wasn't working correctly - it seemed to have some problems fetching the file, either caused by CORS or some other reason. On the other hand the embed option worked (cause i can see below it doesn't anymore) flawlesly without messing with the generated code anymore. Yes, it is somewhat less efficient - the data is stored and fetched as text (whereas with preload it should be fetched as binary) but the size of our dataset isn't staggering so it doesn't cause much of a slowdown. In short it's more comfortable and reliable when using the generated file as a react module.

File: build_wasm.sh (hunk 5 above)
Comment (Mexolius) [reply build_wasm.sh:32]: the @ specifies where in the filesystem created by emcc the files would reside. So in this instance it's `../../libdedx/data@data/` which translates as take files from directory `../../libdedx/data` and put them in directory data in the newly created filesystem. It's there since in libdedx the local directory path is data/ and so the dedx_file_access.c expects to find the files in the data dir.

File: src/Backend/WASMWrapper.js (hunk 1 above)
Comment (Mexolius) [reply src/Backend/WASMWrapper.js:1]: Yes.

File: build_wasm.sh (hunk 4 above)
Comment (grzanka) [reply build_wasm.sh:31]:
```suggestion
# remember to use --embed-file instead of --preload-file for web based development
# # Embed- as opposed to preload - embeds the data into the output .js file. 
# They both should have the same effect - the data should be available on the site 
# but the preload option requires the generated .js file to load the file when it sets the module up. 
# This wasn't working correctly - it seemed to have some problems fetching the file, 
# either caused by CORS or some other reason. 
# On the other hand, the embed option worked flawlessly without messing with the generated code anymore. 
# Yes, it is somewhat less efficient - the data is stored and fetched as text 
# (whereas with preload it should be fetched as binary) 
# but the size of our dataset isn't staggering 
# so it doesn't cause much of a slowdown. 
# In short it's more comfortable and reliable when using the generated file as a react module.
```

Comment (grzanka) [reply build_wasm.sh:31]: Better to add that as a comment in the code

File: src/Backend/WASMWrapper.js [hunk 6]
```diff
@@ -1,10 +1,11 @@
+import TraceFactory from './TraceFactory.js'
 import Module from './weblibdedx.js'
 
 export default class WASMWrapper {
-    #_wasm;
+    #_wasm
     #programsSize = 20
-    #ionsSize = 20
-    #materialsSize = 200
+    #ionsSize = 120
```
Comment (grzanka) [inline src/Backend/WASMWrapper.js:7]: explain in the code from where did you get these values

File: src/Backend/WASMWrapper.js [hunk 7]
```diff
@@ -63,26 +62,71 @@ export default class WASMWrapper {
...
 
         return this.getNames(untyped, 'dedx_get_material_name')
     }
 
-    getTrace(program, ion, material) {
-        return {
-            x: [10, 20, 30, 40, 50, 60, 70, 80, 90, 100],
-            y: Array.from(new Array(10), _ => Math.random() * 100)
-        }
+    async getTrace(program, ion, material, method, plot_using){
```
Comment (grzanka) [inline src/Backend/WASMWrapper.js:80]: What do you mean by Trace ?

File: src/Backend/WASMWrapper.js [hunk 8]
```diff
@@ -63,26 +62,71 @@ export default class WASMWrapper {
...
+        if(method === 0)
+            return await this.getTraceByInterval(program,ion,material,plot_using)
+        else
+            return await this.getTraceByPoints(program,ion,material,plot_using)
+    }
+
+    async getTraceByPoints(program, ion, material, points) {
+        const wasm = await this.wasm()
+        const start = wasm.ccall('dedx_get_min_energy', 'number', ['number','number'],[program,ion])
+        const end = wasm.ccall('dedx_get_max_energy', 'number', ['number','number'],[program,ion])
```
Comment (grzanka) [inline src/Backend/WASMWrapper.js:90]: Can you write to the console (or somewhere else) min and max energy values ?
I suspect there is something wrong there

**Review by grzanka — CHANGES_REQUESTED** [review]
see comments

File: src/Backend/WASMWrapper.js (hunk 7 above)
Comment (Mexolius) [reply src/Backend/WASMWrapper.js:80]: Trace is what i decided to call a singe line on the graph. Basically a pair of x,y- arrays of coordinates to draw on the graph enriched with metadata like name and color. I'm open to changing the name but i can't think of anything better.

Comment (grzanka) [reply src/Backend/WASMWrapper.js:80]: What about a name like `DataSerie` ?
In our domain Trace has a more distant meaning.

File: src/Backend/DataSeriesFactory.js [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    }
+
+    static getXAxis(start,end,points,interval){
+        console.log(points)
+        const arr = Array.from(new Array(points-1),(_,key)=>{return (Number)((start + interval * (key+1).toFixed(3)))})
+        arr.push(end)
+        return arr
+    }
+
+    static getStoppingPowerValues(xs, stepFunction){
```
Comment (grzanka) [inline src/Backend/DataSeriesFactory.js:19]: This function is not specific to StoppingPower values, as it may accept whatever function. What about simply calling it `getYValues` ?

File: src/Backend/DataSeriesFactory.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+        const interval = (end-start)/points
+        return this.getXAxis(start,end, points,interval)
+    }
+
+    static getXAxisByInterval(start,end,interval){
+        const points = Math.ceil((end-start)/interval)
+        return this.getXAxis(start,end,points,interval)
+    }
+
+    static getXAxis(start,end,points,interval){
```
Comment (grzanka) [inline src/Backend/DataSeriesFactory.js:12]: What is the purpose of this method ?

File: src/Backend/DataSeriesFactory.js [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+        console.log(points)
+        const arr = Array.from(new Array(points-1),(_,key)=>{return (Number)((start + interval * (key+1).toFixed(3)))})
```
Comment (grzanka) [inline src/Backend/DataSeriesFactory.js:14]: Explain the logic here. I am especially worried about `toFixed(3)` - why is this added ? Won't this lead to a different numerical results than calculations with full precision ?

File: src/Backend/DataSeriesFactory.js (hunk 10 above)
Comment (grzanka) [reply src/Backend/DataSeriesFactory.js:12]: `points` is unclear - is it number of points or list of points ?
If this is number of points, then call it differently.

Also - how this will behave if we have:
 - start  = 0
 - stop = 100
 - points = 20
 - interval = 1
 - 
What if will produce ?

File: src/Backend/WASMWrapper.js [hunk 12]
```diff
@@ -63,26 +67,73 @@ export default class WASMWrapper {
...
+    async getDataSeries(program, ion, material, method, plot_using){
+        if(method === 0)
+            return await this.getDataSeriesByInterval(program,ion,material,plot_using)
+        else
+            return await this.getDataSeriesByPoints(program,ion,material,plot_using)
+    }
+
+    async getDataSeriesByPoints(program, ion, material, points) {
+        const wasm = await this.wasm()
+        const start = wasm.ccall('dedx_get_min_energy', 'number', ['number','number'],[program,ion])
```
Comment (grzanka) [inline src/Backend/WASMWrapper.js:94]: Better to rename start to `min_energy` to have more meaning

File: src/Backend/DataSeriesFactory.js (hunk 10 above)
Comment (Mexolius) [reply src/Backend/DataSeriesFactory.js:12]: It's basically a private method of this class, but react doesn't support private class member methods as they are an experimental feature. Calling the function with given values will produce a series of 20 points - 19 separated by the interval with the last point being the end point. But since it's technically private it should never be called with those params as the other one is calculated aboved based on the other

Comment (Mexolius) [reply src/Backend/DataSeriesFactory.js:12]: And the purpose is to actually generate X axis values as oposed to getXAxisByPoints and getXAxisByInterval which calculate the other parameter and call this function. Maybe we could use '_' to signify that it's private then? And also change the name to "generateXAxisValues"

File: src/Backend/DataSeriesFactory.js (hunk 11 above)
Comment (Mexolius) [reply src/Backend/DataSeriesFactory.js:14]: It was added to prevent the floating point error from being a factor in the stp calc. Sadly it didn't work and i forgot to delete it

File: src/Backend/DataSeriesFactory.js [hunk 13]
```diff
@@ -0,0 +1,29 @@
+export default class TraceFactory {
+    static getXValuesByPoints(min_energy, max_energy, pointNumber) {
+        const interval = (max_energy - min_energy) / (pointNumber - 1)
```
Comment (grzanka) [inline src/Backend/DataSeriesFactory.js:3]: Imagine user requests a single point. Then you will have division by zero.
Simple solution for that is to check if number of points is == 1 and return then (max + min)/2.

File: src/Backend/DataSeriesFactory.js [hunk 14]
```diff
@@ -0,0 +1,29 @@
+export default class TraceFactory {
+    static getXValuesByPoints(min_energy, max_energy, pointNumber) {
```
Comment (grzanka) [inline src/Backend/DataSeriesFactory.js:2]: rename `pointNumber` to `number_of_points`

Comment (grzanka) [reply src/Backend/DataSeriesFactory.js:2]: This function is not specific to energy, I suggest to rename arguments to `min_value` and `max_value`

File: src/Backend/DataSeriesFactory.js [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+        const interval = (max_energy - min_energy) / (pointNumber - 1)
+        return this.getXAxis(min_energy, max_energy, pointNumber, interval)
+    }
+
+    static getXValuesByInterval(min_energy, max_energy, interval) {
+        // I know +0.01 is cheating but it works for steps as small as 0.1 which generates 2.5k points
+        // and it's the simplest solution to the problem of generating x-values much larger than the max_energy
+        // currently we still go above max by min, ie. if max is 250 and min is 0.00025
+        // the last value is 250.00025 - this allows to keep the intervals constant
+        const pointNumber = Math.ceil((max_energy - min_energy + 0.01) / interval)
```
Comment (grzanka) [inline src/Backend/DataSeriesFactory.js:12]: What if we introduce the feature that user requests a smaller interval for energy by hand, for example: 0.02 to 0.03 - what then ?

File: src/Backend/WASMWrapper.js [hunk 16]
```diff
@@ -1,10 +1,16 @@
+import DataSeriesFactory from './DataSeriesFactory.js'
 import Module from './weblibdedx.js'
 
 export default class WASMWrapper {
-    #_wasm;
+    #_wasm
+
+    // The values used here are based on array sizes in dedx_program_const.h
+    // They aren't one-to-one mapping but there's a gurentee they are always
```
Comment (grzanka) [inline src/Backend/WASMWrapper.js:8]:
```suggestion
    // They aren't one-to-one mapping but there's a guarantee they are always 
```

File: src/Backend/WASMWrapper.js [hunk 17]
```diff
@@ -1,10 +1,16 @@
+import DataSeriesFactory from './DataSeriesFactory.js'
 import Module from './weblibdedx.js'
 
 export default class WASMWrapper {
-    #_wasm;
+    #_wasm
+
+    // The values used here are based on array sizes in dedx_program_const.h
+    // They aren't one-to-one mapping but there's a gurentee they are always 
+    // equal to or bigger than the expected number of entities they describe
```
Comment (grzanka) [inline src/Backend/WASMWrapper.js:9]:
```suggestion
    // greater than or equal to the expected number of entities they describe
```

File: src/Backend/DataSeriesFactory.js [hunk 18]
```diff
@@ -0,0 +1,24 @@
+export default class TraceFactory {
+    static getXValuesByPoints(minValue, maxValue, numberOfPoints) {
+        if(numberOfPoints == 1) return [(maxValue - minValue)/2]
```
Comment (grzanka) [inline src/Backend/DataSeriesFactory.js:3]:
```suggestion
        if(numberOfPoints == 1) return [(maxValue + minValue)/2.0]
```

mean value is calculated as `(a+b)/2`  not as `(a-b)/2`

File: src/Backend/DataSeriesFactory.js [hunk 19]
```diff
@@ -0,0 +1,24 @@
+export default class TraceFactory {
+    static getXValuesByPoints(minValue, maxValue, numberOfPoints) {
+        if(numberOfPoints === 1) return [(maxValue - minValue)/2]
```
Comment (grzanka) [inline src/Backend/DataSeriesFactory.js:3]:
```suggestion
        if(numberOfPoints === 1) return [(maxValue + minValue)/2.0]
```


mean value is calculated as (a+b)/2 not as (a-b)/2

**Review by grzanka — APPROVED** [review]
LGTM
